Spigot-Server-Patches/0334-Adding-alternate-and-more-even-spawn-mechanic.patch

```diff
@@ -0,0 +1,547 @@
+@@ -16,18 +16,35 @@ public enum EnumCreatureType {
+         this.h = flag1;
+     }
+ 
++    // Paper start
++    public static EnumCreatureType getCreatureType(Entity entity) {
++        Class<?> clazz = entity.getClass();
++
++        if(MONSTER.getInnerClass().isAssignableFrom(clazz)) return MONSTER;
++        if(CREATURE.getInnerClass().isAssignableFrom(clazz)) return CREATURE;
++        if(AMBIENT.getInnerClass().isAssignableFrom(clazz)) return AMBIENT;
++        if(WATER_CREATURE.getInnerClass().isAssignableFrom(clazz)) return WATER_CREATURE;
++
++        return null;
++    }
++    // Paper end
++
++    public Class<? extends IAnimal> getInnerClass() { return a(); } // Paper - OBFHELPER
+     public Class<? extends IAnimal> a() {
+         return this.e;
+     }
+ 
++    public int getSpawnLimit() { return b(); } // Paper - OBFHELPER
+     public int b() {
+         return this.f;
+     }
+ 
++    public boolean isPassive() { return c(); } // Paper - OBFHELPER
+     public boolean c() {
+         return this.g;
+     }
+ 
++    public boolean isRare() { return d(); } // Paper - OBFHELPER
+     public boolean d() {
+         return this.h;
+     }
+diff --git a/src/main/java/net/minecraft/server/PlayerChunk.java b/src/main/java/net/minecraft/server/PlayerChunk.java
+index fcd9f5491..6f04d4759 100644
+--- a/src/main/java/net/minecraft/server/PlayerChunk.java
++++ b/src/main/java/net/minecraft/server/PlayerChunk.java
+@@ -249,6 +249,7 @@ public class PlayerChunk {
+         return false;
+     }
+ 
++    public boolean isDone() { return e(); } // Paper - OBFHELPER
+     public boolean e() {
+         return this.done;
+     }
+diff --git a/src/main/java/net/minecraft/server/SpawnerCreature.java b/src/main/java/net/minecraft/server/SpawnerCreature.java
+index f525fd1b4..23a0fc6a3 100644
+--- a/src/main/java/net/minecraft/server/SpawnerCreature.java
++++ b/src/main/java/net/minecraft/server/SpawnerCreature.java
+@@ -1,10 +1,12 @@
+ package net.minecraft.server;
+ 
+-import com.google.common.collect.Sets;
+ import java.util.Iterator;
+ import java.util.List;
+ import java.util.Random;
+ import java.util.Set;
++import java.util.HashSet; // Paper
++import java.util.Map;
++import java.util.EnumMap; // Paper
+ import javax.annotation.Nullable;
+ import org.apache.logging.log4j.LogManager;
+ import org.apache.logging.log4j.Logger;
+@@ -21,10 +23,288 @@ public final class SpawnerCreature {
+     private static final Logger a = LogManager.getLogger();
+     private static final int b = (int) Math.pow(17.0D, 2.0D);
+     private final LongHashSet c = new LongHashSet(); // CraftBukkit
++    public static int seed = "paper".hashCode(); // Paper
+ 
+     public SpawnerCreature() {}
+ 
++    public static class ChunkCoordOrdinalInt3Tuple extends ChunkCoordIntPair {
++        public int ordinal;
++        public int cachedHashCode = -1;
++
++        public static int mix(int hash, int data) {
++            int k = data;
++
++            k *= 0xcc9e2d51;
```

Comment:
What are all these magic numbers? Mind adding a comment or explaining them here? 😛

Reply:
this is all from the Murmur3 32-bit algorthim. Should maybe add a link, but I did leave a comment with the algorithm name at the top of that chunk

Reply:
and for context: its a fast and evenly distributed (but non-cryptographic) 32b hashing algo (its used in scala, stole it from there)

Reply:
Ah sorry, didn't see it.

Reply:
ill leave a little blurb about what and why. got so spoiled by scala having this algo out-of-the-box in scala, just got used to em